File: Level Designer Evolved/src/Edge.java
// ======================================= //
// Level Designer Evolved                  //
//                                         //
// Author: Kevin Scroggins                 //
// E-Mail: nitro404@hotmail.com            //
// ======================================= //

import java.util.StringTokenizer;
import java.io.*;
import java.awt.*;

/**
 * The Edge class represents a pair of Vertices which form a segmented line.
 * 
 * @author Kevin Scroggins
 */
public class Edge {
	
	/** The first Vertex of the Edge. */
	public Vertex a;
	
	/** The second Vertex of the Edge. */
	public Vertex b;
	
	/**
	 * Constructs an Edge based on the two specified Vertices.
	 * 
	 * @param a the first Vertex.
	 * @param b the second Vertex.
	 */
	public Edge(Vertex a, Vertex b) {
		this.a = a;
		this.b = b;
	}
	
	/**
	 * Returns the second Vertex of the Edge.
	 * 
	 * @return the second Vertex of the Edge.
	 */
	public Vertex getPointA() {
		return this.a;
	}
	
	/**
	 * Returns the first Vertex of the Edge.
	 * 
	 * @return the first Vertex of the Edge.
	 */
	public Vertex getPointB() {
		return this.b;
	}
	
	/**
	 * Checks to see if the specified Vertex is a member of the Edge.
	 * 
	 * @param v the Vertex to check.
	 * @return true if the specified Vertex is a member of the Edge.
	 */
	public boolean containsVertex(Vertex v) {
		return this.a == v || this.b == v;
	}
	
	/**
	 * Computes the vertical distance between the two Vertices and returns it.
	 * 
	 * @return the vertical distance between the two Vertices.
	 */
	public double getDeltaX() {
		return this.b.x - this.a.x;
	}

	/**
	 * Computes the horizontal distance between the two Vertices and returns it.
	 * 
	 * @return the horizontal distance between the two Vertices.
	 */
	public double getDeltaY() {
		return this.b.y - this.a.y;
	}
	
	/**
	 * Creates an Edge from the specified String and returns it.
	 * 
	 * Parses the Edge from a String of the form:
	 * "x1, y1; x2, y2" where
	 * x1 and y1 are the x and y coordinates of the first Vertex of the Edge and
	 * x2 and y2 are the x and y coordinates of the second Vertex of the Edge.
	 * 
	 * @param input the String to parse the Edge from.
	 * @return the Edge parsed from the String.
	 */
	public static Edge parseFrom(String input) {
		if(input == null || input.trim().length() == 0) {
			return null;
		}
		
		String data = input.trim();
		
		// extract the strings representing each vertex, then parse and store them
		StringTokenizer st = new StringTokenizer(data, ";, ", false);
		Vertex a = Vertex.parseFrom(st.nextToken() + ", " + st.nextToken());
		Vertex b = Vertex.parseFrom(st.nextToken() + ", " + st.nextToken());
		
		return new Edge(a, b);
	}
	
	/**
	 * Writes the Edge to the specified PrintWriter.
	 * 
	 * Outputs the Edge to the form:
	 * "x1, y1; x2, y2" where
	 * x1 and y1 are the x and y coordinates of the first Vertex of the Edge and
	 * x2 and y2 are the x and y coordinates of the second Vertex of the Edge.
	 * 
	 * @param out the PrintWriter to write the Edge to.
	 * @throws IOException if there was an error writing to the output stream.
	 */
	public void writeTo(PrintWriter out) throws IOException {
		a.writeTo(out);
		out.print("; ");
		b.writeTo(out);
	}
	
	/**
	 * Renders the Edge onto the specified Graphics object.
	 * 
	 * @param g the Graphics object to render the Edge onto.
	 */
	public void paintOn(Graphics g) {
		g.drawLine(a.x, a.y, b.x, b.y);
	}
	
	/* (non-Javadoc)
	 * @see java.lang.Object#equals(java.lang.Object)
	 */
	public boolean equals(Object o) {
		if(o == null || !(o instanceof Edge)) {
			return false;
		}
		
		Edge e = (Edge) o;
		
		// check to see that each Vertex in the Edge matches the corresponding Vertex of the other
		return this.a == e.a && this.b == e.b;
	}
	
	/* (non-Javadoc)
	 * @see java.lang.Object#toString()
	 */
	public String toString() {
		// return a String representation of the Edge in the form (x1, y1) (x2, y2)
		return "(" + this.a.x + ", " + this.a.y + ") (" + this.b.x + ", " + this.b.y + ")";
	}
	
}

File: Firecracker Engine/Source/Graph/Edge.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.IO;

namespace Firecracker_Engine {

	public class Edge {

		private Vertex m_a;
		private Vertex m_b;

		public Edge(Vertex a, Vertex b) {
			this.a = a;
			this.b = b;
		}
		
		public Vertex a {
			get { return m_a; }
			set { if(value != null) { m_a = value; } }
		}

		public Vertex b {
			get { return m_b; }
			set { if (value != null) { m_b = value; } }
		}

		public float deltaX {
			get { return m_b.x - m_a.x; }
		}

		public float deltaY {
			get { return m_b.y - m_a.y; }
		}

		public float length {
			get { return (float) Math.Sqrt(Math.Pow(m_b.x - m_a.x, 2) + Math.Pow(m_b.y - m_a.y, 2)); }
		}

		public bool containsVertex(Vertex v) {
			return v != null && (m_a.Equals(v) || m_b.Equals(v));
		}

		public static Edge parseFrom(String input) {
			if(input == null) { return null; }

			String data = input.Trim();

			if(data.Length == 0) { return null; }

			string[] vertexData = data.Split(';');

			if(vertexData.Length != 2) { return null; }

			Vertex a = Vertex.parseFrom(vertexData[0]);
			Vertex b = Vertex.parseFrom(vertexData[1]);

			if(a == null || b == null) { return null; }

			return new Edge(a, b);
		}

		public void writeTo(StreamWriter output) {
			if(output == null) { return; }

			a.writeTo(output);
			output.Write("; ");
			b.writeTo(output);
		}

		public override int GetHashCode() {
			return base.GetHashCode();
		}

		public override bool Equals(object o) {
			Edge e = o as Edge;
			return e != null && m_a.Equals(e.m_a) && m_b.Equals(e.m_b);
		}

		public override string ToString() {
			return a + " " + b;
		}

	}

}
